File: asian/python/utils/verification.py
import numpy as np
import pandas as pd

def verify_arithmetic_parity(S0_list, sigma_list, call_MC, put_MC, call_FD, put_FD, r, T, K):
    """
    Verifies the arithmetic Asian put–call parity and returns a DataFrame.
    
    The theoretical parity (at t=0) is given by:
      Call - Put = exp(-rT)*(((exp(rT)-1)/(rT))*S0 - K)
    """
    parity_results = []
    for i, s0 in enumerate(S0_list):
        parity_theory = np.exp(-r * T) * ((((np.exp(r * T) - 1) / (r * T)) * s0) - K)
        for j, sigma in enumerate(sigma_list):
            lhs_MC = call_MC[i, j] - put_MC[i, j]
            lhs_FD = call_FD[i, j] - put_FD[i, j]
            err_MC = lhs_MC - parity_theory
            err_FD = lhs_FD - parity_theory
            parity_results.append([s0, sigma, lhs_MC, lhs_FD, parity_theory, err_MC, err_FD])
    parity_results = np.array(parity_results)
    colNames = ['S0', 'sigma', '(Call-Put)MC', '(Call-Put)FD', 'Theory', 'ErrorMC', 'ErrorFD']
    parityTable = pd.DataFrame(parity_results, columns=colNames)
    return parityTable

File: cpp/asian/utils/verification.cpp
#include "verification.h"
#include <cmath>
#include <sstream>
#include <iomanip>
#include <numeric>

// Function to compute the parity table.
std::vector<ParityRecord> verifyArithmeticParity(
    const std::vector<double>& S0_list,
    const std::vector<double>& sigma_list,
    const std::vector<std::vector<double>>& call_MC,
    const std::vector<std::vector<double>>& put_MC,
    const std::vector<std::vector<double>>& call_FD,
    const std::vector<std::vector<double>>& put_FD,
    double r, double T, double K
) {
    std::vector<ParityRecord> table;
    // Loop over each S0 value.
    for (size_t i = 0; i < S0_list.size(); i++) {
        double s0 = S0_list[i];
        // Compute the theoretical parity:
        //    exp(-rT)*(((exp(rT)-1)/(rT))*s0 - K)
        double parity_theory = std::exp(-r * T) * ( ((std::exp(r * T) - 1) / (r * T)) * s0 - K );
        // Loop over each sigma value.
        for (size_t j = 0; j < sigma_list.size(); j++) {
            ParityRecord rec;
            rec.s0 = s0;
            rec.sigma = sigma_list[j];
            double lhs_MC = call_MC[i][j] - put_MC[i][j];
            double lhs_FD = call_FD[i][j] - put_FD[i][j];
            rec.callPutMC = lhs_MC;
            rec.callPutFD = lhs_FD;
            rec.theory    = parity_theory;
            rec.errorMC   = lhs_MC - parity_theory;
            rec.errorFD   = lhs_FD - parity_theory;
            table.push_back(rec);
        }
    }
    return table;
}

std::string parityTableToString(const std::vector<ParityRecord>& table) {
    std::ostringstream ss;

    // We won't set a global format at the start,
    // because we want to set it individually for each column.

    // Column headers (you can keep them at a fixed width).
    ss << "Arithmetic Put-Call Parity Comparison for Asian Options\n";
    ss << std::setw(12) << "S0"
       << std::setw(12) << "sigma"
       << std::setw(15) << "(Call-Put)MC"
       << std::setw(15) << "(Call-Put)FD"
       << std::setw(15) << "Theory"
       << std::setw(15) << "ErrorMC"
       << std::setw(15) << "ErrorFD"
       << "\n";

    // Separator line
    ss << std::string(12 + 12 + 15 + 15 + 15 + 15 + 15, '-') << "\n";

    for (const auto& rec : table) {
        // 1) Print S0 in fixed, say 2 decimal places
        ss << std::fixed << std::setprecision(2)
           << std::setw(12) << rec.s0;

        // 2) Print sigma in fixed, say 2 decimal places
        ss << std::fixed << std::setprecision(2)
           << std::setw(12) << rec.sigma;

        // 3) Print the rest in scientific, e.g. 6 decimal places
        ss << std::scientific << std::setprecision(6)
           << std::setw(15) << rec.callPutMC
           << std::setw(15) << rec.callPutFD
           << std::setw(15) << rec.theory
           << std::setw(15) << rec.errorMC
           << std::setw(15) << rec.errorFD
           << "\n";
    }
    return ss.str();
}

